File: go/cpu/csvparse.go
package main

import (
	"fmt"
	"os"
	"slices"
	"strconv"
	"strings"

	"github.com/mohamedmenasy/go-vs-rust-benchmarks/go/internal/common"
	"github.com/mohamedmenasy/go-vs-rust-benchmarks/go/internal/harness"
)

// csvparse: parse an in-memory request log (ts,user_id,endpoint,status,
// latency_ms,bytes) and aggregate per endpoint. The file is read (and, for
// Rust, UTF-8 validated) during setup; the timed work is scanning lines and
// fields, parsing integers and decimals, and updating a string-keyed map.
type csvAgg struct {
	count             uint64
	latSum, latMax    float64
	bytes, n4xx, n5xx uint64
}

type csvParse struct{ data string }

func mustInt(s string) int64 {
	v, err := strconv.ParseInt(s, 10, 64)
	if err != nil {
		panic(err)
	}
	return v
}

func csvAggregate(data string) uint64 {
	aggs := make(map[string]*csvAgg)
	rows := uint64(0)
	_, rest, _ := strings.Cut(data, "\n") // header
	for len(rest) > 0 {
		var line string
		line, rest, _ = strings.Cut(rest, "\n")
		if line == "" {
			continue
		}
		tsS, r, _ := strings.Cut(line, ",")
		userS, r, _ := strings.Cut(r, ",")
		ep, r, _ := strings.Cut(r, ",")
		statusS, r, _ := strings.Cut(r, ",")
		latS, bytesS, _ := strings.Cut(r, ",")
		_ = mustInt(tsS)
		_ = mustInt(userS)
		status := mustInt(statusS)
		lat, err := strconv.ParseFloat(latS, 64)
		if err != nil {
			panic(err)
		}
		size := mustInt(bytesS)
		a := aggs[ep]
		if a == nil {
			a = &csvAgg{}
			aggs[ep] = a
		}
		a.count++
		a.latSum += lat
		if lat > a.latMax {
			a.latMax = lat
		}
		a.bytes += uint64(size)
		if status >= 400 && status < 500 {
			a.n4xx++
		} else if status >= 500 {
			a.n5xx++
		}
		rows++
	}
	keys := make([]string, 0, len(aggs))
	for k := range aggs {
		keys = append(keys, k)
	}
	slices.Sort(keys)
	d := common.NewDigest()
	for _, k := range keys {
		a := aggs[k]
		d.AddString(k)
		d.Add(a.count)
		d.AddF64(a.latSum)
		d.AddF64(a.latMax)
		d.Add(a.bytes)
		d.Add(a.n4xx)
		d.Add(a.n5xx)
	}
	d.Add(rows)
	return d.Sum()
}

func (w *csvParse) Run() uint64 { return csvAggregate(w.data) }

func setupCSVParse(p harness.Params) (harness.Instance, harness.Work, error) {
	path := p.MustStr("input")
	b, err := os.ReadFile(path)
	if err != nil {
		return nil, harness.Work{}, fmt.Errorf("read %s: %w", path, err)
	}
	return &csvParse{data: string(b)}, harness.Work{Unit: "bytes", PerRun: float64(len(b)), InputBytes: int64(len(b))}, nil
}

File: rust/cpu/src/csvparse.rs
//! Parse an in-memory request log (ts,user_id,endpoint,status,latency_ms,
//! bytes) and aggregate per endpoint (mirror of go/cpu/csvparse.go). The file
//! is read and UTF-8 validated during setup; the timed work is scanning lines
//! and fields, parsing integers and decimals, and updating a string-keyed map
//! (std HashMap with its default SipHash-1-3 hasher: the baseline).

use std::collections::HashMap;

use bench_common::{Digest, Instance, Params, Work};

#[derive(Default)]
struct Agg {
    count: u64,
    lat_sum: f64,
    lat_max: f64,
    bytes: u64,
    n4xx: u64,
    n5xx: u64,
}

fn int(s: &str) -> i64 {
    s.parse().expect("integer field")
}

pub fn aggregate(data: &str) -> u64 {
    let mut aggs: HashMap<String, Agg> = HashMap::new();
    let mut rows = 0u64;
    let rest = data.split_once('\n').map_or("", |(_, r)| r); // header
    for line in rest.split('\n') {
        if line.is_empty() {
            continue;
        }
        let mut f = line.split(',');
        let (ts, user, ep, status, lat, size) = (
            f.next().unwrap(),
            f.next().unwrap(),
            f.next().unwrap(),
            f.next().unwrap(),
            f.next().unwrap(),
            f.next().unwrap(),
        );
        let _ = int(ts);
        let _ = int(user);
        let status = int(status);
        let lat: f64 = lat.parse().expect("decimal field");
        let size = int(size);
        let update = |a: &mut Agg| {
            a.count += 1;
            a.lat_sum += lat;
            if lat > a.lat_max {
                a.lat_max = lat;
            }
            a.bytes += size as u64;
            if (400..500).contains(&status) {
                a.n4xx += 1;
            } else if status >= 500 {
                a.n5xx += 1;
            }
        };
        // One hash lookup on the hit path, like Go's `aggs[ep]`; the key is
        // only allocated the first time an endpoint is seen.
        if let Some(a) = aggs.get_mut(ep) {
            update(a);
        } else {
            let mut a = Agg::default();
            update(&mut a);
            aggs.insert(ep.to_string(), a);
        }
        rows += 1;
    }
    let mut keys: Vec<&String> = aggs.keys().collect();
    keys.sort();
    let mut d = Digest::new();
    for k in keys {
        let a = &aggs[k];
        d.add_str(k);
        d.add(a.count);
        d.add_f64(a.lat_sum);
        d.add_f64(a.lat_max);
        d.add(a.bytes);
        d.add(a.n4xx);
        d.add(a.n5xx);
    }
    d.add(rows);
    d.sum()
}

struct CsvParse {
    data: String,
}

impl Instance for CsvParse {
    fn run(&mut self) -> u64 {
        aggregate(&self.data)
    }
}

pub fn setup(p: &Params) -> Result<(Box<dyn Instance>, Work), String> {
    let path = p.must_str("input");
    let bytes = std::fs::read(&path).map_err(|e| format!("read {path}: {e}"))?;
    let n = bytes.len();
    let data = String::from_utf8(bytes).map_err(|e| format!("{path}: {e}"))?;
    Ok((Box::new(CsvParse { data }), Work { unit: "bytes", per_run: n as f64, input_bytes: n as u64 }))
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::testutil::{golden_cpu, params, run_once};
    use bench_common::golden_hex;

    #[test]
    fn matches_golden() {
        let c = &golden_cpu()["csvparse"];
        let path = bench_common::repo_root().join(c["fixture"].as_str().unwrap());
        let (mut inst, _) = setup(&params(&[("input", path.display().to_string())])).unwrap();
        assert_eq!(run_once(&mut inst).0, golden_hex(&c["digest"]));
    }
}
